src/Passes/Source/Commandline/ConfigBind.hpp

```diff
@@ -0,0 +1,207 @@
+        ~ConfigBind() override              = default;
+
+        /// Constructor to bind value to parameter. This class holds a reference to a variable together
+        /// with the name it is expected to have when passed through the parameter parser.
+        ConfigBind(Type& bind, T default_value, String const& name, String const& description);
+
+        // Interface implementation
+        //
+
+        /// Adds the argument to the parser.
+        bool setupArguments(ParameterParser& parser) override;
+
+        /// Configures the bound value. This method examines the parsed input
+        /// and use updates the bound value accordingly.
+        bool configure(ParameterParser const& parser) override;
+
+        /// String representation of the bound value.
+        String value() override;
+
+      private:
+        /// Generic function to setup arguments of any type.
+        template <typename R> bool setupArguments(ParameterParser&, R const&);
+
+        /// Specialised function setting arguments up for booleans.
+        bool setupArguments(ParameterParser& parser, bool const&);
+
+        /// Generic function that changes the parameter name based on the value type and default value.
+        template <typename R> void alterNameBasedOnType(R const& default_value);
+
+        /// Specialised function that changes the parameter name based on default value for booleans.
+        void alterNameBasedOnType(bool const& default_value);
+
+        /// Generic string serialization.
+        template <typename A> String valueAsString(A const&);
+
+        /// Specialised serialization for booleans.
+        template <typename A> String valueAsString(EnableIf<A, bool, A> const&);
+
+        /// Generic deserialization of string values from parser.
+        template <typename R> void loadValue(ParameterParser const& parser, R const& default_value);
+
+        /// Specialised deserialization of string values from parser for booleans.
+        template <typename A> void loadValue(ParameterParser const& parser, EnableIf<A, bool, A> const& default_value);
+
+        /// Specialised deserialization of string values from parser for strings.
+        template <typename A>
+        void loadValue(ParameterParser const& parser, EnableIf<A, String, A> const& default_value);
+
+        Type& bind_;          ///< Bound variable to be updated.
+        Type  default_value_; ///< Default value.
+    };
+
+    template <typename T>
+    ConfigBind<T>::ConfigBind(Type& bind, T default_value, String const& name, String const& description)
+      : IConfigBind(name, description)
+      , bind_{bind}
+      , default_value_{std::move(default_value)}
+    {
+        alterNameBasedOnType(default_value_);
+    }
+
+    template <typename T> template <typename R> void ConfigBind<T>::alterNameBasedOnType(R const& default_value)
+    {
+        std::stringstream ss{""};
+        ss << default_value;
+        setDefault(ss.str());
```

Comment:
The `ss.str()` returns `std::string()`. The param of `setDefault()` has type `microsoft::quantum::IConfigBind::String`.

This line (106) tells that the parameter of `setDefault()` has to be based 
* either on `std::string` rather than `microsoft::quantum::IConfigBind::String` (even though currently the latter of them is an alias of the former of them);
* or on a type the `std::string` can be implicitly cast to.

Consider changing the signature of `setDefault()` to 
`void IConfigBind::setDefault(std::string const& v)` from 
`void IConfigBind::setDefault(String const& v)`.
And if you do, then based on the fragment
```c++
    void IConfigBind::setDefault(String const& v)
    {
        str_default_value_ = v;
    }
```
consider changing the type of `str_default_value_` also to `std::string`.

Reply:
I've added a static cast to `String` all relevant places. As long as `String` is a type alias this is noop which is optimised away and if we, for instance, would change `String` to `llvm::StringRef` or a `string_view` then it is converted in accordance with the available conversion paths.